### ISSUE TITLE
Secret credentials has an incorrect `instance_name` (when externalName is used)

### ISSUE THREAD

**Issue by pvnn** [SAP/sap-btp-service-operator#270]
Hi colleagues,

**We face an issue when creating a `ServiceInstance` with `externalName`, which according to the documentaton is the name for the service instance in SAP BTP and is what applications want to actually consume in their apps.**

**Steps to repoduce:**

1. Create a service instance as shown below:
``` yaml
    apiVersion: services.cloud.sap.com/v1
    kind: ServiceInstance
    metadata:
        name: my-service-instance
    spec:
        serviceOfferingName: sample-service
        servicePlanName: sample-plan
        externalName: my-service-instance-external
        parameters:
          key1: val1
          key2: val2
```

2. Create a service binding:
``` yaml
  apiVersion: services.cloud.sap.com/v1
  kind: ServiceBinding
  metadata:
      name: my-binding
  spec:
      serviceInstanceName: my-service-instance
      externalName: my-binding-external
      secretName: mySecret
      parameters:
        key1: val1
        key2: val2      
```
3. Check the secret details in the k8s cluster where this is executed.
```
kubectl get secrets mySecret
```
 
**Issue:**
The secret metadata always contains the instance name as the k8s service instance name. **This is incorrect when an externalName is specified as the consumers (apps) expect to consume this e.g. via [@sap/xsenv](https://www.npmjs.com/package/@sap/xsenv) using instance name from BTP (e.g. as present in the service catalog)**

```
apiVersion: v1
type: Opaque
kind: Secret
metadata:
  name: mySecret
stringData:
  type: sample-service
  provider: SAP
  instance_guid: xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
  instance_name: my-service-instance # This should be my-service-instance-external
  ...
```

Issue seems to be in the [service binding controller code here](https://github.com/SAP/sap-btp-service-operator/blob/main/controllers/servicebinding_controller.go#L702)
IMO updating this to use externalName from instance (when it exists should solve the issue).

Please take over and provide a fix for this.

Thanks & Regards,
Pavan

**Comment by kerenlahav** [comment]
Hi Pavan,
I don't think it's a bug, **external** name is the name that is being used outside of the cluster, it will be the instance name in Service Manager and the name that you will see in cockpit and SM api, but within the cluster the instance name is metadata.name
Why do you need to use the external name within the cluster?
Thanks,
Keren

**Comment by pvnn** [comment]
Hi Keren,
As mentioned in my original description.. when the credentials are consumed say by a Pod within the cluster and an app wants to access the BTP service credentials via instance name - it would do so using the instance name in SM api and/or BTP cockpit and NOT via the k8s cluster metadata.name. Using the cluster specific name in the credential metadata is a bug as consuming apps do not care what the k8s instance name of the service is.. they need to know the actual instance name from SM / BTP cockpit.
The fix I proposed addresses that by updating the OSB credential metadata with the instance_name as present in BTP cockpit.

**Comment by kerenlahav** [comment]
some applications might require the k8s name and some the BTP name, I will add another key to the secret "instance_external_name" that will contain the external name

**Comment by pvnn** [comment]
Hi Keren,
That will not help as all consuming libraries need to adopt and support this new metadata. E.g. CAP, xsenv etc.. How should we/consumers now handle that with your proposal? (we do not want to create multiple mounts/secrets here to work around the issue).
See for instance: https://www.npmjs.com/package/@sap/xsenv#usage-in-kubernetes: as to how one must mount this in k8s.

Shouldn't your guidance instead be that **applications that rely on k8s name** - do not set any externalName? With that existing apps should work as they did before (i.e. use the same name for k8s and BTP service instance).

Regards,
Pavan

**Comment by pvnn** [comment]
Hi,
Sorry the new property you introduce does not solve issue (as mentioned above). It instead causes all consumers to adopt and support this new metadata.

We think the instance_name property in the Secret metadata, never worked as intended and hence fixing it would be the proper solution here.

**Comment by loewenstein** [comment]
Actually, changing this would be incompatible, service tags might be a better option in general for selecting service credentials and why would users actually not use the instance name they created?

Adding external name as a property seems to be the best way forward, wouldn't it?

**Comment by pvnn** [comment]
> ...why would users actually not use the instance name they created?

Exactly.. but if you try to consider the actual consumers here (who are trying to consume the BTP service credentials), the _instance_name_ that one would expect to see in the service metadata would be the **actual** service instance name (from BTP) and not the "k8s instance name". 
Especially if you want to put yourselved in the shoes of app developers.. they would wish to use the BTP service instance name (e.g. from BTP cockpit) - just like say they would when they'd create this manually -or- via _any other mechanism_ on CF/K8s.

> Adding external name as a property seems to be the best way forward, wouldn't it?

No. As mentioned before.. most of the consumers CAP, [@sap/xsenv](https://www.npmjs.com/package/@sap/xsenv#usage-in-kubernetes), any other library/consumers who have their existing way of consuming BTP service credentials - do not know what this "external name" is. To be frank, this would be something that is just invented here by the BTP service operator, which is NOT the only way to create service instances in BTP.


**Moreover.. another discrepancy here that we mentioned already to the developers:** 
if **_instance_name_** is really meant for K8s instance name then shouldn’t **_instance_guid_** also point to the UId of the K8s instance(?).
In our opinion however it makes no sense do that, instead we should always point to the **actual** BTP service _instance_name_ and _instance_guid_ when it exists (e.g. when _externalName_ is set). For existing consumers, who do not wish this behaviour they could just (continue to) not set any _externalName_ and get the behaviour you seem to always mention (k8s instance name = BTP instance name = metadata credentials instance_name).

**Basically the service metadata added to the secret by this operator was always incorrect and needs to be fixed.. even if it may be incompatible. Or do we stop fixing every issue claiming it's incompatible?**

**Comment by avilupu** [comment]
Hi. After re-evaluation, we accepted your argument that this is a bug.
A fix is delivered with [release v0.5.3]https://github.com/SAP/sap-btp-service-operator/releases/tag/v0.5.3 version.
In the created secret, the value of the instance name key is the BTP name, which is maintained in the external name property of the binding, not in the metadata name.

**Comment by Pavan-SAP** [comment]
@avilupu Thanks for looking into this.
 
A small typo/issue in your comment above:
"**In the created secret, the value of the instance_name key is the BTP name, which is maintained in the external name property of the ~binding~ instance (instance_name contains the external service instance name as in BTP when present), not the metadata name.**"

The change (#338) itself looks good though and there intance (external) name is used in secret `instance_name` property. 👍
